dhall/tests/Dhall/Test/QuickCheck.hs

```diff
@@ -457,16 +458,20 @@ instance Arbitrary URL where
 
         authority <- arbitrary
 
-        path <- arbitrary
-
-        let validQueryCharacter c =
+        let validPChar c =
                    ('\x41' <= c && c <= '\x5A')
                 || ('\x61' <= c && c <= '\x7A')
                 || ('\x30' <= c && c <= '\x39')
-                || c `elem` ("-._~!$&'()*+,;=:@/?" :: String)
+                || c `elem` ("-._~!$&'()*+,;=:@" :: String)
+
+        let validPath File{ directory = Directory{ components }, file } =
+                all (Text.all validPChar) (file : components)
+
+        path <- suchThat arbitrary validPath
+
+        let validQueryCharacter c = validPChar c || c `elem` ("/?" :: String)
 
-        let validQuery  Nothing  = True
-            validQuery (Just q ) = Text.all validQueryCharacter q
+        let validQuery = all (Text.all validQueryCharacter)
```

Comment:
It's not directly related to #1934, but shouldn't we also generate percent-encoded characters here?

Reply:
@sjakobi: Adding one more format/parse cycle seems to me like an undesirable workaround for bugs in the `Arbitrary` instances, because:

* The work-around is imperfect and does not guarantee that you actually fix the problem
* The work-around can introduce new issues by masking bugs

The reason why is that adding an additional preprocessing step of formatting and parsing the expression can lead to an expression which is entirely different in character from the original expression.

To illustrate this, suppose there were a pretty-printing bug where we forget to escape variables with spaces in their name using backticks.  In other words, what should have been pretty-printed as:

```dhall
`a b`
```

... is instead pretty-printed as:

```
a b
```

If you were to add one more format/parse cycle to the formatting code it would incorrectly detect that there was no formatting issue because the first time it would as `App "a" "b"`, which is a valid but different expression, and then it would find not issue because formatting `a b` is idempotent.

More generally, I try to abide by what I call the "golden rule of software engineering", which is that engineers should prefer to push fixes upstream instead of working around problems downstream.  I don't mind fixing these `Arbitrary` instances because they were indeed broken and QuickCheck was correct to point that out.  Yes, they can sometimes be difficult to implement correctly, but that's the essential complexity behind property testing.

Reply:
Yeah, good points. It sounds like we'll end up reproducing many of the parser rules for the generators. Maybe that would be simplified by encoding the invariants in the AST types, and for example use different `File` types for HTTP imports and absolute imports.

I also wonder whether, once we switch the parser to `Earley`, we could somehow use its [generator capabilities](http://hackage.haskell.org/package/Earley-0.13.0.1/docs/Text-Earley-Generator.html) to generate test cases…

Reply:
Yeah, I to think there is value in using separate `File` types now that the grammar for URLs and file paths no longer matches